### PR TITLE
added opencart under store. to deployment

### PR DESCRIPTION
**Pull request by ghost**
here my attempt to add opencart under store.domain.tld - please review

### CODE REVIEW

File: roles/mailserver/files/docker-compose.yml [hunk 1]
```diff
@@ -138,9 +153,19 @@ services:
       - mailserver
       - postfixadmin
       - rainloop
+{% if www_is_opencart %}
+      - /mnt/docker/opencart:/opencart
+{% endif %}
```
Comment (ksylvan) [inline roles/mailserver/files/docker-compose.yml]: Is this set of lines in the wrong spot?
```yaml
{% if www_is_opencart %}
       - /mnt/docker/opencart:/opencart
{% endif %}
```

needs to be under volumes.

Does the nginx container even need to have a volume with visibility into the opencart container?

File: roles/mailserver/files/docker-compose.yml [hunk 2]
```diff
@@ -138,9 +153,19 @@ services:
...
+      - postfixadmin
+      - rainloop
 {% if www_is_wordpress %}
       - wordpress
 {% else %}
+{% if www_is_opencart %}
+      - opencart
+{% else %}
       - contact
 {% endif %}
```
Comment (ksylvan) [inline roles/mailserver/files/docker-compose.yml]: The if/else logic here does not seem right. 

Looks like it should be something along the lines of:
```yaml
{% if www_is_wordpress %}
    - wordpress
{% elif www_is_opencart %}
    - opencart
{% else %}
    - contact
{% endif %}
```
see http://jinja.pocoo.org/docs/2.10/templates/#builtin-filters

File: roles/mailserver/tasks/main.yml [hunk 3]
```diff
@@ -146,6 +146,39 @@
...
+    register: php_custom
+  - name: Set up empty custom.ini for php modifications
+    file:
+      path: /mnt/docker/php/custom.ini
+      state: touch
+      owner: 82
+      group: 82
+    when: not php_custom.stat.exists
+  when: www_is_opencart
+
```
Comment (ksylvan) [inline roles/mailserver/tasks/main.yml]: The block above seems like just a cut/paste of the wordpress stanzas. It also repeats the php `custom.ini` setup. Is the PHP `custom.ini` even needed for `opencart`? It is its own self-contained container.

I could be wrong, but I think the only needed piece here is 

```yaml
- name: Set up opencart directory
   file:
     path: /mnt/docker/opencart
     state: directory
     recurse: yes
    owner: 82
    group: 82
- name: Set up opencart nginx site
   template:
     src: files/nginx/www-opencart.conf
     dest: /mnt/docker/nginx/sites-enabled/opencart.conf
     owner: 991
     group: 991
   notify:
   - restart nginx
   - restart opencart
```

**Comment by ksylvan** [comment]
I added some comments. Have you tested this?

You can do `make save` (save your current setup), then `make reset`

Then you can spin up a VPS, set up your control host's `/etc/hosts` to have something like:
```
test.dev mail.test.dev www.test.dev postfixadmin.test.dev webmail.test.dev spam.test.dev 123.45.67.89
```
(replace that IP with the real IP) and run `make`

Use `test.dev` as the domain and answer `true` for the "Use self-signed certs" question. Then you can try it out with browsing to `www.test.dev`

**Review by ksylvan — COMMENTED** [review]
Look at and address the above. Thank you for your work!

**Comment by outbackdingo** [comment]
Ive made the subtle changes suggested then i tested it, it appears to work
up until however the only container i dont see running is nginx

docker ps
CONTAINER ID        IMAGE                            COMMAND
                 CREATED             STATUS              PORTS


                                NAMES

27135f9ed762        hardware/rainloop                "run.sh"
                2 minutes ago       Up 2 minutes        8888/tcp


                                rainloop

6eb445adae9d        hardware/postfixadmin            "tini -- run.sh"
        2 minutes ago       Up 2 minutes        8888/tcp

                                postfixadmin

bb16e0cb0615        hardware/mailserver:1.1-latest   "run.sh"
                2 minutes ago       Up 2 minutes        0.0.0.0:25->25/tcp,
0.0.0.0
:143->143/tcp, 0.0.0.0:587->587/tcp, 0.0.0.0:993->993/tcp, 465/tcp,
0.0.0.0:4190->4190/tcp, 11334/tcp   mailserver
5f9183d4369c        bitnami/opencart:3               "/app-entrypoint.s..."
  2 minutes ago       Up 2 minutes        80/tcp, 443/tcp

                                opencart

5bfa9fcf770c        mariadb:10.1                     "docker-entrypoint..."
  2 minutes ago       Up 2 minutes        3306/tcp

                                mariadb

8670d1ee5542        redis:3.2-alpine                 "docker-entrypoint..."
  2 minutes ago       Up 2 minutes        6379/tcp

                                redis

heres the output from the run
TASK [mailserver : Install the ufw-docker-hack script]
******************************************************************************************

ok: [mail.test.dev]

TASK [mailserver : Install ufw-with-docker action for fail2ban]
*********************************************************************************

ok: [mail.test.dev]

TASK [mailserver : fail2ban filter for bots that hit the php apps (301)]
************************************************************************
ok: [mail.test.dev]

TASK [mailserver : fail2ban setup for apache_badbots and bots-redirect]
*************************************************************************
ok: [mail.test.dev]

TASK [mailserver : Start stack if needed]
*******************************************************************************************************

fatal: [mail.test.dev]: FAILED! => {"changed": true, "cmd":
"/usr/local/bin/docker-compose up -d", "delta": "0:00:04.294178", "end":
"2017-11-09
12:11:34.414386", "failed": true, "msg": "non-zero return code", "rc": 1,
"start": "2017-11-09 12:11:30.120208", "stderr": "Creating redis ... \r
\nCreating mariadb ... \r\nCreating redis\nCreating
mariadb\n\u001b[1A\u001b[2K\rCreating redis ...
\u001b[32mdone\u001b[0m\r\u001b[1B\u001b[1A\u
001b[2K\rCreating mariadb ... \u001b[32mdone\u001b[0m\r\u001b[1BCreating
opencart ... \r\nCreating mailserver ... \r\nCreating opencart\nCreating
mailserver\n\u001b[1A\u001b[2K\rCreating opencart ...
\u001b[32mdone\u001b[0m\r\u001b[1B\u001b[1A\u001b[2K\rCreating mailserver
... \u001b[32mdo
ne\u001b[0m\r\u001b[1BCreating contact ... \r\nCreating postfixadmin ...
\r\nCreating rainloop ... \r\nCreating contact\nCreating postfixadmin\nC
reating rainloop\n\u001b[1A\u001b[2K\rCreating contact ...
\u001b[31merror\u001b[0m\r\u001b[1B\nERROR: for contact  Cannot start
service contact:
oci runtime error: container_linux.go:265: starting container process
caused \"process_linux.go:368: container init caused \\\"rootfs_linux.go:5
7: mounting \\\\\\\"/mnt/docker/contact/index.php\\\\\\\" to rootfs
\\\\\\\"/var/lib/docker/overlay2/78998b54a2ee3d886de8f7db12bbf1db5aa511244ce3
9adca3775ef5bc81106d/merged\\\\\\\" at
\\\\\\\"/var/lib/docker/overlay2/78998b54a2ee3d886de8f7db12bbf1db5aa511244ce39adca3775ef5bc81106d/merged/c
ontact/index.php\\\\\\\" caused \\\\\\\"not a directory\\\\\\\"\\\"\"\n:
Are you trying to mount a directory onto a file (or vice-versa)? Check i
f the specified host path exists and is the expected
type\n\u001b[1A\u001b[2K\rCreating rainloop ...
\u001b[32mdone\u001b[0m\r\u001b[1B\u001b[1A\
u001b[2K\rCreating postfixadmin ...
\u001b[32mdone\u001b[0m\r\u001b[1B\nERROR: for contact  Cannot start
service contact: oci runtime error: cont
ainer_linux.go:265: starting container process caused
\"process_linux.go:368: container init caused \\\"rootfs_linux.go:57:
mounting \\\\\\\"/mnt
/docker/contact/index.php\\\\\\\" to rootfs
\\\\\\\"/var/lib/docker/overlay2/78998b54a2ee3d886de8f7db12bbf1db5aa511244ce39adca3775ef5bc81106d/mer
ged\\\\\\\" at
\\\\\\\"/var/lib/docker/overlay2/78998b54a2ee3d886de8f7db12bbf1db5aa511244ce39adca3775ef5bc81106d/merged/contact/index.php\\\\\\\"
caused \\\\\\\"not a directory\\\\\\\"\\\"\"\n: Are you trying to mount a
directory onto a file (or vice-versa)? Check if the specified host pat
h exists and is the expected type\nEncountered errors while bringing up the
project.", "stderr_lines": ["Creating redis ... ", "Creating mariadb
... ", "Creating redis", "Creating mariadb", "\u001b[1A\u001b[2K",
"Creating redis ... \u001b[32mdone\u001b[0m",
"\u001b[1B\u001b[1A\u001b[2K", "
Creating mariadb ... \u001b[32mdone\u001b[0m", "\u001b[1BCreating opencart
... ", "Creating mailserver ... ", "Creating opencart", "Creating mail
server", "\u001b[1A\u001b[2K", "Creating opencart ...
\u001b[32mdone\u001b[0m", "\u001b[1B\u001b[1A\u001b[2K", "Creating
mailserver ... \u001b[32
mdone\u001b[0m", "\u001b[1BCreating contact ... ", "Creating postfixadmin
... ", "Creating rainloop ... ", "Creating contact", "Creating postfixa
dmin", "Creating rainloop", "\u001b[1A\u001b[2K", "Creating contact ...
\u001b[31merror\u001b[0m", "\u001b[1B", "ERROR: for contact  Cannot start
service contact: oci runtime error: container_linux.go:265: starting
container process caused \"process_linux.go:368: container init caused \\\"
rootfs_linux.go:57: mounting \\\\\\\"/mnt/docker/contact/index.php\\\\\\\"
to rootfs \\\\\\\"/var/lib/docker/overlay2/78998b54a2ee3d886de8f7db12b
bf1db5aa511244ce39adca3775ef5bc81106d/merged\\\\\\\" at
\\\\\\\"/var/lib/docker/overlay2/78998b54a2ee3d886de8f7db12bbf1db5aa511244ce39adca3775ef5
bc81106d/merged/contact/index.php\\\\\\\" caused \\\\\\\"not a
directory\\\\\\\"\\\"\"", ": Are you trying to mount a directory onto a
file (or v
ice-versa)? Check if the specified host path exists and is the expected
type", "\u001b[1A\u001b[2K", "Creating rainloop ... \u001b[32mdone\u001b[
0m", "\u001b[1B\u001b[1A\u001b[2K", "Creating postfixadmin ...
\u001b[32mdone\u001b[0m", "\u001b[1B", "ERROR: for contact  Cannot start
service c
ontact: oci runtime error: container_linux.go:265: starting container
process caused \"process_linux.go:368: container init caused \\\"rootfs_lin
ux.go:57: mounting \\\\\\\"/mnt/docker/contact/index.php\\\\\\\" to rootfs
\\\\\\\"/var/lib/docker/overlay2/78998b54a2ee3d886de8f7db12bbf1db5aa51
1244ce39adca3775ef5bc81106d/merged\\\\\\\" at
\\\\\\\"/var/lib/docker/overlay2/78998b54a2ee3d886de8f7db12bbf1db5aa511244ce39adca3775ef5bc81106d/m
erged/contact/index.php\\\\\\\" caused \\\\\\\"not a
directory\\\\\\\"\\\"\"", ": Are you trying to mount a directory onto a
file (or vice-versa)
? Check if the specified host path exists and is the expected type",
"Encountered errors while bringing up the project."], "stdout": "", "stdout_
lines": []}



RUNNING HANDLER [mailserver : restart nginx]
****************************************************************************************************


RUNNING HANDLER [mailserver : restart mailserver]
***********************************************************************************************

       to retry, use: --limit
@/home/dingo/docker-mail-server-forked/mailserver.retry

PLAY RECAP
**************************************************************************************************************************************

mail.test.dev              : ok=39   changed=9    unreachable=0    failed=1

**Comment by ksylvan** [comment]
Hmmm... Look in your test server in /mnt/docker/nginx/logs/* and see if you can figure out why.

**Comment by ksylvan** [comment]
It also occurs to me that if you want the `opencart` container to show up as store.yourdomain.tld, then it does not actually conflict with the existing setup where www is either a standard nginx setup (with files and ~user settings, or a simple contact form) or wordpress.

So we'll need to re-work that logic a bit.

**Comment by ksylvan** [comment]
So we can probably just have a `use_opencart` variable and in that case we should add `store.{{ domain_name }}` to the set of hosts that are included in the SSL cert we generate from lets encrypt.

**Comment by ghost** [comment]
no logs.... ill push the changes... i do see that the bitnami cart came up, so ill ask is there a way on the server i can try to kickstart the nginx container and see if its going south for some reason ? i even tried a docker logs -f nginx ... which said it didnt exist

 ls -al /mnt/docker/nginx/log/
total 8
drwxr-xr-x 2 991 991 4096 Nov  9 12:10 .
drwxr-xr-x 5 991 991 4096 Nov  9 12:10 ..
-rw-r--r-- 1 991 991    0 Nov  9 12:10 access.log

**Comment by ghost** [comment]
and logically i would simply like to add the opencart to the deployment, as i have other www pages for the primary website not based on wordpress either to go into www. While thinking about it i could also deploy quassel-core in a container, but ill deal with that down the road.....  right now www web site and opencart is my target

File: roles/mailserver/files/docker-compose.yml [hunk 4]
```diff
@@ -94,6 +94,34 @@ services:
...
+    container_name: contact
+    # restart: always
+    volumes:
+      - /mnt/docker/contact/config:/contact/config
+      - /mnt/docker/contact/ssmtp:/etc/ssmtp
+      - /mnt/docker/contact/index.php:/contact/index.php
+    depends_on:
+      - mailserver
+{% endif %}
+
```
Comment (ksylvan) [inline roles/mailserver/files/docker-compose.yml]: The whole else-endif block is using the wrong level of indentation. This whole block should not show up as a change at all at this point.

File: roles/mailserver/files/docker-compose.yml [hunk 5]
```diff
@@ -138,6 +166,13 @@ services:
       - mailserver
       - postfixadmin
       - rainloop
+{% if use_opencart %}
+      - /mnt/docker/opencart:/opencart
```
Comment (ksylvan) [inline roles/mailserver/files/docker-compose.yml]: This is not right. This is probably why the nginx container is not getting created.  This section is for cross-container dependency for startup, not volume links.

File: roles/secure-server/tasks/main.yml [hunk 6]
```diff
@@ -74,24 +74,10 @@
               state=present
   notify: Restart ssh
 
-- name: Disallow password authentication
```
Comment (ksylvan) [inline roles/secure-server/tasks/main.yml]: Undo these changes. I want to disallow password authentication and root logins into the server. After these playbooks run, the only way into the server (except via the console, which many VPS providers can provide) is to ssh using the 'deploy' user, from the control host.

You can add other ssh keys manually to `~deploy/.ssh/authorized_keys` in order to login from other accounts that you have.

**Review by ksylvan — COMMENTED** [review]
I've reviewed this again. I think I know why nginx is not coming up for you.

File: roles/mailserver/files/docker-compose.yml (hunk 5 above)
Comment (ghost) [reply roles/mailserver/files/docker-compose.yml]: containers coming up now, there were some config issues with the opencart.conf for nginx.... though its not pretty but i can see the cart now.

File: roles/secure-server/tasks/main.yml (hunk 6 above)
Comment (ghost) [reply roles/secure-server/tasks/main.yml]: got it.... forgot i did that.....

**Review by ksylvan — COMMENTED** [review]
Did you re-push your latest changes?

Also, now that I think about it, the `use_opencart` variable default might need to be moved to a different level (instead of sitting in the `mailserver` role) because it will also affect the list of hosts that we include in the LE SSL initial creation.

File: roles/mailserver/files/docker-compose.yml (hunk 4 above)
Comment (ksylvan) [reply roles/mailserver/files/docker-compose.yml]: This still needs to be fixed back to what it was originally.

**Comment by ksylvan** [comment]
Cleaned up the changes a bit. This still does not account for adding the "store.{{ domain_name }}" into the generated SSL cert, though.

**Comment by ghost** [comment]
ok seems to still work, however the cart at store.test.dev looks wrong, not displaying properly, however it does come up. everything else appears correct ill test with wordpress, with regular www files, make sure they all works properly, not sure what you mean by This still does not account for adding the "store.{{ domain_name }}" into the generated SSL cert, though.

**Comment by ghost** [comment]
ok these all appear to do the proper things in the proper order if selected / not selected..... Ive tested it everyway possible. the storefront still doesnt display right... i left it up if you want to take a peak
148.251.24.6    test.dev mail.test.dev www.test.dev postfixadmin.test.dev webmail.test.dev spam.test.dev store.test.dev

**Comment by ksylvan** [comment]
Looking at it now... I'm getting Connection Refused with that IP address.

**Comment by ghost** [comment]
hah i was messing with it so its restored now if you dint get a chance to see it..... its back up, i noticed when adding various web templates also when pushing them it will fail .... ? does it not just copy whats ever is in www/test.dev/files to the web  space

**Comment by ksylvan** [comment]
I brought up my latest tweaks on your testing branch locally. When I visit `store.test.dev`, I see a weirdly formatted store page in google chrome. On the top right of the address bar, I see a little icon, "This page is trying to load scripts from unauthenticated sources" and then I click on the icon and click on "load unsafe scripts" to get a nicely formatted page.

Also needed to ensure that http and https both get proxy_passsed to the opencart container. Pull it and try again.

**Comment by ksylvan** [comment]
The last piece to polish this work is to automatically set up the SMTP settings for the `opencart` container.
Reference: https://github.com/bitnami/bitnami-docker-opencart#smtp-configuration
I think you can put `mailserver` as the SMTP host, and the mail setup already accepts relaying from any of the containers running in the local docker network.

**Comment by ghost** [comment]
ok this appears to work fine now any ideas about the web space issue failing when transferring some files ?? I think thats the last issue im seeing really......

**Comment by ghost** [comment]
changed: [mail.test.dev] => (item=./files/revolution/js/extensions/revolution.extension.migration.min.js)
changed: [mail.test.dev] => (item=./files/revolution/js/extensions/revolution.extension.video.min.js)
changed: [mail.test.dev] => (item=./files/revolution/js/extensions/revolution.extension.layeranimation.min.js)
failed: [mail.test.dev] (item=./files/revolution/js/jquery.themepunch.tools.min.js) => {"changed": false, "failed": true, "item": "./files/revolution/js/jquery.themepunch.tools.min.js", "msg": "AnsibleUndefinedVariable: 'LT' is undefined"}
changed: [mail.test.dev] => (item=./files/revolution/js/source/jquery.themepunch.revolution.js)
failed: [mail.test.dev] (item=./files/revolution/js/source/jquery.themepunch.tools.min.js) => {"changed": false, "failed": true, "item": "./files/revolution/js/source/jquery.themepunch.tools.min.js", "msg": "AnsibleUndefinedVariable: 'LT' is undefined"}
changed: [mail.test.dev] => (item=./files/revolution/js/source/jquery.themepunch.enablelog.js)
changed: [mail.test.dev] => (item=./files/revolution/js/jquery.themepunch.enablelog.js)
changed: [mail.test.dev] => (item=./files/revolution/js/jquery.themepunch.revolution.min.js)
changed: [mail.test.dev] => (item=./files/revolution/fonts/pe-icon-7-stroke/css/pe-icon-7-stroke.css)


specifically this file....

**Comment by ghost** [comment]
changed: [mail.test.dev] => (item=./people/scottk/portfolio-item.html)
changed: [mail.test.dev] => (item=./people/scottk/service-item.html)

RUNNING HANDLER [mailserver : restart nginx] ****************************************************************************************************

RUNNING HANDLER [mailserver : restart mailserver] ***********************************************************************************************

RUNNING HANDLER [mailserver : restart opencart] *************************************************************************************************
        to retry, use: --limit @/home/dingo/docker-mail-server-forked/mailserver.retry

PLAY RECAP **************************************************************************************************************************************
mail.test.dev              : ok=39   changed=10   unreachable=0    failed=1

**Comment by ksylvan** [comment]
```
(item=./files/revolution/js/extensions/revolution.extension.layeranimation.min.js)
failed: [mail.test.dev] (item=./files/revolution/js/jquery.themepunch.tools.min.js) => {"changed": false, "failed": true, "item": "./files/revolution/js/jquery.themepunch.tools.min.js", "msg": "AnsibleUndefinedVariable: 'LT' is undefined"}
changed: [mail.test.dev] => (item=./files/revolution/js/source/jquery.themepunch.revolution.js)
failed: [mail.test.dev] (item=./files/revolution/js/source/jquery.themepunch.tools.min.js) => {"changed": false, "failed": true, "item": "./files/revolution/js/source/jquery.themepunch.tools.min.js", "msg": "AnsibleUndefinedVariable: 'LT' is undefined"}
```

My guess is there’s something in those files that look like variables. Or maybe `{{ something }}` that
the template module tries to substitute. Might need to fix the way files are copied or templated.
My initial idea was that non binary files would get templated (so you could use variables
like `domain_name` in the files) but that obviously might not work for JavaScript files.

**Comment by ghost** [comment]
maybe just copy **.js insttead of trying to manipulate them, right now i
have to do it manually after i put a index.html in www/test.dev/files

im also looking at / optionally to populate the
Unbound: recursive caching DNS resolver with DNSSEC support
NSD : authoritative DNS server with DNSSEC support

which I also dont see being installed via the ansible run, however they
exist now on the https://github.com/hardware/mailserver

ill look a tha after this pull request is completed and see if we can get
those in the mix as well. probably also try to see if theres a way to
manage them via the web

**Comment by ghost** [comment]
i did find this so i guess i can look at integrating it into a container
nginx/php/mysql and the pages
http://www.lissyara.su/articles/freebsd/programms/nsdadmin/

**Comment by ghost** [comment]
wow too easy to deploy ...   already up
https://github.com/obi12341/docker-pdns

**Comment by ksylvan** [comment]
```
maybe just copy **.js insttead of trying to manipulate them, right now i
have to do it manually after i put a index.html in www/test.dev/files
```
Please add an issue for me to take care of that.

**Comment by ksylvan** [comment]
Squashed and merged. Feel free to delete your testing branch, re-pull the main develop or master branch, and provide any additional feedback or fixes for the opencart setup. I added the SMTP settings.
